packages/byods/src/data-source-service/index.ts

```diff
@@ -0,0 +1,133 @@
+/* eslint-disable no-useless-catch */
+/* eslint-disable no-console */
+/* eslint-disable class-methods-use-this */
+/* eslint-disable import/no-extraneous-dependencies */
+
+import {v4 as uuidv4} from 'uuid';
+import {decodeJwt, JWTPayload} from 'jose';
+import DataSourceClient from '../data-source-client/index';
+
+export default class DataSourceService {
+  private dataSourceClient: DataSourceClient;
+  private timer: any;
+
+  /**
+   * Creates an instance of DataSourceService.
+   * @param {DataSourceClient} dataSourceClient - The  dataSourceClient instance to use for DataSourceService.
+   * @example
+   * const dataSourceService = new DataSourceService(dataSourceClient);
+   */
+
+  constructor(dataSourceClient: DataSourceClient) {
+    this.dataSourceClient = dataSourceClient;
+  }
+
+  /**
+   * This method refreshes the DataSource token using dataSourceId and tokenLifetimeMinute
+   * @param {string} dataSourceId The id of data source
+   * @param {number} tokenLifetimeMinutes The Life time minutes for the data source. <=1440
+   * @returns {Promise<void>}
+   */
+
+  public async refreshDataSourceToken(
+    dataSourceId: string,
+    tokenLifetimeMinutes?: number
+  ): Promise<void> {
+    try {
+      if (dataSourceId) {
+        if (tokenLifetimeMinutes) {
+          this.startAutoRefresh(dataSourceId, tokenLifetimeMinutes - 5);
```

Comment:
This should be a percentage instead of -5.
Make the percentage random from ranging from 5% to 10% of the given tokenLifeTimesMinutes

Reply:
@pagour98 , Also make sure that the final percentage amounts to at least 5mins before expiry

Reply:
**_Good Catch_** @bharath063 
**_This has been addressed!_**